massa-module-cache/src/controller.rs

```diff
@@ -144,21 +161,37 @@ impl ModuleCache {
                 }
             }
         };
-        Ok(module)
+        Ok((module, execution_gas))
     }
 
-    /// Load a temporary module from arbitrary bytecode
+    /// Load a temporary module from arbitrary bytecode.
+    /// Pays the Singlepass compilation cost.
+    /// Also checks that the provided execution gas is enough to pay for the instance creation cost.
+    ///
+    /// Returns the module and the remaining gas after compilation.
     pub fn load_tmp_module(
         &self,
         bytecode: &[u8],
         limit: u64,
-    ) -> Result<RuntimeModule, CacheError> {
+    ) -> Result<(RuntimeModule, u64), CacheError> {
         debug!("load_tmp_module");
-        Ok(RuntimeModule::new(
+        let remaining = limit
+            .checked_sub(self.cfg.gas_costs.sp_compilation_cost)
+            .ok_or(CacheError::LoadError(
+                "Not enough gas to pay compilation".to_string(),
+            ))?;
+        remaining
+            .checked_sub(self.cfg.gas_costs.max_instance_cost)
+            .ok_or(CacheError::LoadError(
+                "Provided max gas is below the default instance creation cost".to_string(),
+            ))?;
+
+        let module = RuntimeModule::new(
             bytecode,
-            limit,
+            remaining,
```

Comment:
In previous uses of `RuntimeModule::new()` the value used for the `limit` parameter is `self.cfg.gas_costs.max_instance_cost`.
Here we have already subtracted this value to the parameter we are providing. It sounds weird to me.
I agree that `load_tmp_module` should return `remaining` (as it's computed here) but I think that we have to call `new()` with `self.cfg.gas_costs.max_instance_cost` (hence the comment above, the limit param can be dropped)